File: lib/domain/models/pending_action.dart
enum PendingActionType {
  disableProtection,
  removeBlocklistDomain,
  changeWebhookUrl,
  increaseSensitiveActionDelay,
  decreaseSensitiveActionDelay,
  deactivateDeviceAdmin,
}

enum PendingActionState { pending, applied, cancelled }

class PendingAction {
  const PendingAction({
    required this.id,
    required this.type,
    required this.requestedAt,
    required this.readyAt,
    required this.state,
    this.payload = const {},
  });

  final String id;
  final PendingActionType type;
  final DateTime requestedAt;
  final DateTime readyAt;
  final PendingActionState state;
  final Map<String, String> payload;

  bool get isReadyToApply =>
      state == PendingActionState.pending &&
      DateTime.now().isAfter(readyAt);

  Duration get timeRemaining {
    final remaining = readyAt.difference(DateTime.now());
    return remaining.isNegative ? Duration.zero : remaining;
  }

  PendingAction copyWith({PendingActionState? state}) {
    return PendingAction(
      id: id,
      type: type,
      requestedAt: requestedAt,
      readyAt: readyAt,
      state: state ?? this.state,
      payload: payload,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'type': type.name,
        'requestedAt': requestedAt.toIso8601String(),
        'readyAt': readyAt.toIso8601String(),
        'state': state.name,
        'payload': payload,
      };

  factory PendingAction.fromJson(Map<String, dynamic> json) {
    return PendingAction(
      id: json['id'] as String,
      type: PendingActionType.values.byName(json['type'] as String),
      requestedAt: DateTime.parse(json['requestedAt'] as String),
      readyAt: DateTime.parse(json['readyAt'] as String),
      state: PendingActionState.values.byName(json['state'] as String),
      payload: Map<String, String>.from(json['payload'] as Map? ?? {}),
    );
  }
}

File: desktop/src/domain/pending_action.cpp
#include "pending_action.h"

#include <algorithm>
#include <stdexcept>

namespace clearguard::domain {

bool PendingAction::isReadyToApply() const {
    return state == PendingActionState::Pending && std::chrono::system_clock::now() >= readyAt;
}

std::chrono::seconds PendingAction::timeRemaining() const {
    auto remaining = std::chrono::duration_cast<std::chrono::seconds>(readyAt - std::chrono::system_clock::now());
    return std::max(remaining, std::chrono::seconds::zero());
}

std::string toString(PendingActionType type) {
    switch (type) {
        case PendingActionType::DisableProtection: return "disableProtection";
        case PendingActionType::RemoveBlocklistDomain: return "removeBlocklistDomain";
        case PendingActionType::ChangeWebhookUrl: return "changeWebhookUrl";
        case PendingActionType::ChangeRemoteBlocklistUrl: return "changeRemoteBlocklistUrl";
        case PendingActionType::IncreaseSensitiveActionDelay: return "increaseSensitiveActionDelay";
        case PendingActionType::DecreaseSensitiveActionDelay: return "decreaseSensitiveActionDelay";
    }
    throw std::invalid_argument("Unknown PendingActionType");
}

PendingActionType pendingActionTypeFromString(const std::string &value) {
    if (value == "disableProtection") return PendingActionType::DisableProtection;
    if (value == "removeBlocklistDomain") return PendingActionType::RemoveBlocklistDomain;
    if (value == "changeWebhookUrl") return PendingActionType::ChangeWebhookUrl;
    if (value == "changeRemoteBlocklistUrl") return PendingActionType::ChangeRemoteBlocklistUrl;
    if (value == "increaseSensitiveActionDelay") return PendingActionType::IncreaseSensitiveActionDelay;
    if (value == "decreaseSensitiveActionDelay") return PendingActionType::DecreaseSensitiveActionDelay;
    throw std::invalid_argument("Unknown PendingActionType: " + value);
}

std::string toString(PendingActionState state) {
    switch (state) {
        case PendingActionState::Pending: return "pending";
        case PendingActionState::Applied: return "applied";
        case PendingActionState::Cancelled: return "cancelled";
    }
    throw std::invalid_argument("Unknown PendingActionState");
}

PendingActionState pendingActionStateFromString(const std::string &value) {
    if (value == "pending") return PendingActionState::Pending;
    if (value == "applied") return PendingActionState::Applied;
    if (value == "cancelled") return PendingActionState::Cancelled;
    throw std::invalid_argument("Unknown PendingActionState: " + value);
}

}
